src/nupic/encoders/ScalarEncoder.cpp

```diff
@@ -21,115 +23,175 @@
  */
 
 /** @file
- * Implementations of the ScalarEncoder and PeriodicScalarEncoder
+ * Implementation of the ScalarEncoder
  */
 
-#include <algorithm> //std::fill
-#include <cmath>
-
+#include <algorithm> // std::min
+#include <numeric>   // std::iota
+#include <math.h>    // isnan
 #include <nupic/encoders/ScalarEncoder.hpp>
 
 namespace nupic {
 
-std::vector<UInt> ScalarEncoderBase::encode(Real input) {
-    std::vector<UInt> output(getOutputWidth());
-    encodeIntoArray(input, output.data());
-    return output;
-}
+ScalarEncoder::ScalarEncoder(ScalarEncoderParameters &parameters)
+  { initialize( parameters ); }
+
+void ScalarEncoder::initialize(ScalarEncoderParameters &parameters)
+{
+  // Check parameters
+  NTA_CHECK( parameters.minimum < parameters.maximum );
+
+  UInt num_active_args = 0;
+  if( parameters.active     > 0)    num_active_args++;
+  if( parameters.sparsity   > 0.0f) num_active_args++;
+  NTA_CHECK( num_active_args != 0u )
```

Comment:
please use the actual conditions in the check, so rather 
`NTA_CHECK(params.active > 0 || params.sparsity > 0) << "active or sparse must be specified.."

Reply:
On second thought, I'd like to keep what's here now because the current way has a different error message for the two different ways the user could mess this up:
* Neither argument given
* Both arguments given

Reply:
>  two different ways the user could mess this up

then the check should be ==1;

Reply:
> the check should be ==1

Yes.  There are two checks, one for too few arguments ( != 0 ) and one for too many arguments ( == 1 )